# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E - менее 30 000'
        elif 30001 <= income <= 50000:
            return 'D - от 30 000 до 50 000'
        elif 50001 <= income <= 200000:
            return 'C - от 50 000 до 200 000'
        elif 200001 <= income <= 1000000:
            return 'B - от 200 000 до 1 000 000'
        elif income >= 1000001:
            return 'A - больше 1 000 000'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Сгруппируем данные по столбцу **children** и посчитаем процент должников для каждого значения. Для этого разделим количество должников (подсчитанное методом **sum**, т.к. единица в столбце **debt** указывает на наличие задолженности) на общее количество заемщиков (подсчитанное методом **count**) и умножим на 100. 

Объединим эти данные в одну таблицу методом **concat** и переименуем столбцы для более удобного представления. Для наглядности используем метод **display**.

In [22]:
child_debt_sum = data.groupby('children')['debt'].sum()
child_debt_count = data.groupby('children')['debt'].count()
child_data_debt_percent = child_debt_sum / child_debt_count * 100

case_children = pd.concat([child_data_debt_percent, child_debt_count, child_debt_sum], axis=1)
case_children.columns = ['percentage_of_debts','count','sum']
display(case_children)

percentage_of_debts  count   sum
children                                  
0                    7.543822  14091  1063
1                    9.234609   4808   444
2                    9.454191   2052   194
3                    8.181818    330    27
4                    9.756098     41     4
5                    0.000000      9     0

**Вывод:** 
Общее количество заемщиков больше всего в группе без детей, но при этом самый низкий процент должников.

Можно видеть, что чем больше детей - тем меньше люди берут кредит в принципе, однако у тех, что берут вероятность вернуть во время меньше, чем у группы без детей. У группы с 5 детьми отсутствуют задолженности, но девять таких случаев - это не показатель, её можно исключить.

В целом, наличие детей уменьшает вероятность возврата кредита в срок, т.к. процент должников в группах с детьми больше. Можно сделать вывод, что чем больше детей - тем больше процент должников, но в третьей группе он ниже, чем у остальных, на это нужно обратить внимание. 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Возьмем за основу порядок действий из предыдущего исследования. Используя группировку по столбцу **family_status** для столбца **debt**, найдем общее число заемщиков **(count)** и количество должников **(sum)**. Разделив число должников на общее число заемщиков, получим процент должников в каждой группе.

Объединим эти данные в одну таблицу методом concat и переименуем столбцы для более удобного представления. Для наглядности используем метод **display** и отсортируем по столбцу с процентом должников по возрастанию.

In [23]:
family_debt_sum = data.groupby('family_status')['debt'].sum()
family_debt_count = data.groupby('family_status')['debt'].count()
family_data_debt_percent = family_debt_sum / family_debt_count * 100

case_family = pd.concat([family_data_debt_percent, family_debt_count, family_debt_sum], axis=1)
case_family.columns = ['percentage_of_debts','count','sum']
display(case_family.sort_values('percentage_of_debts'))

percentage_of_debts  count  sum
family_status                                         
вдовец / вдова                    6.624606    951   63
в разводе                         7.064760   1189   84
женат / замужем                   7.560558  12261  927
гражданский брак                  9.313014   4134  385
Не женат / не замужем             9.763948   2796  273

**Вывод:** 
Наименьший процент должников среди тех, кто потерял семью (вдовец/вдова). Они же реже всех остальных берут кредиты. Можно предположить, что такие люди, в виду одиночества, проявляют бОльшую осторожность и ответственность при взаимодействии с долговыми обязательствами.

Больше всего берут кредиты женатые пары. Процент должников средний. Вполне логичный показатель, семьям, в особенности молодым, требуются ресурсы на обустройство и развитие, но в виду обстоятельств, не все могут закрыть долги во время. Тем не менее, это не самая большая группа должников, следовательно, наличие семьи способствует возврату кредита вовремя.

Самый большой процент должников в группе тех, кто не был женат/замужем. Но общее число заемщиков в этой группе не самое большое.

Опираясь на результаты предыдущего исследования, можно сделать вывод, что меньше всего вероятность вернуть кредит у тех, кто имеет детей (причем, чем больше - тем хуже), и при этом никогда не был в браке.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Возьмем два столбца исходных данных (**total_income** и **debt**). Создадим отдельную таблицу **case_income_data_categorized**, отсортированную по столбцу дохода и применим функцию **categorize_income** из 16 задания, чтобы создать в ней столбец **total_income_category** с категорями уровня дохода. В функции дополнены описания к категориям, чтобы можно было видеть числовые диапазоны в выходных данных.

Теперь можно применить шаги из предыдущих исследований - при помощи группировки по новому столбцу **total_income_category** найти общее число заемщиков **(income_debt_count)**, количество должников **(income_debt_sum)** и процент тех, кто имеет задолженность **(income_data_debt_percent)**.

Выходные данные отсортируем по проценту должников и выведем результат **case_income**.

In [24]:
case_income_data = data[['total_income','debt']]

case_income_data_categorized = case_income_data.sort_values(by='total_income').reset_index(drop=True)
case_income_data_categorized['total_income_category'] = case_income_data['total_income'].apply(categorize_income)

income_debt_sum = case_income_data_categorized.groupby('total_income_category')['debt'].sum()
income_debt_count = case_income_data_categorized.groupby('total_income_category')['debt'].count()
income_data_debt_percent = income_debt_sum / income_debt_count * 100


case_income = pd.concat([income_data_debt_percent, income_debt_count, income_debt_sum], axis=1)
case_income.columns = ['percentage_of_debts','count','sum']
display(case_income.sort_values('percentage_of_debts'))


percentage_of_debts  count   sum
total_income_category                                        
C - от 50 000 до 200 000                8.094756  15788  1278
D - от 30 000 до 50 000                 8.139535    344    28
B - от 200 000 до 1 000 000             8.202338   4962   407
E - менее 30 000                        9.090909     22     2
A - больше 1 000 000                   12.000000     25     3

**Вывод:** 
Две крайние категории (A, у кого доход менее 30 000 и D, у кого доход более 1 000 000) имеют наибольший процент должников. Однако, общее количество заемщиков в этих группах крайне мало, выводы по ним нельзя считать показательными. Все логично - люди с крайне низким доходом не хотят брать на себя лишние долговые обязательства, а людям с крайне высоким доходом кредиты не нужны.
По категории E тоже недостаточно данных, чтобы считать её значимой, будем делать вывод на основании оставшихся категорий.

Больше всего кредитов берут люди из средней категории C (от 50 000 до 200 000). В этой же категории наименьшее количество должников. Можно предположить, что она является предпочтительной при оценке. Примерно такой же процент должников и в картегории B, но число заемщиков там в три раза меньше. На основании этого, можно сделать вывод, что в данном исследовании вероятность вернуть кредит в срок не зависит от уровня дохода. Однако, стоит учитывать и простую логику - чем выше доход, тем меньше берут кредиты и тем проще их вернуть в срок.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Воспользуемся схемой действий из предыдущего исследования. Возьмем два столбца (**purpose** и **debt**), но будем применять функцию **categorize_purpose** из задания 18 к столбцу **purpose**, чтобы создать новый столбец с категориями по целям (**purpose_category**).

Далее при помощи группировки по новому столбцу **purpose_category** найдем общее число заемщиков (**purpose_debt_count**), количество должников (**purpose_debt_sum**) и процент тех, кто имеет задолженность (**purpose_data_debt_percent**).

Выходные данные отсортируем по проценту должников и выведем результат **case_purpose**.

In [25]:
case_purpose_data = data[['purpose','debt']]

case_purpose_data_categorized = case_purpose_data.sort_values(by='purpose').reset_index(drop=True)
case_purpose_data_categorized['purpose_category'] = case_purpose_data['purpose'].apply(categorize_purpose)

purpose_debt_sum = case_purpose_data_categorized.groupby('purpose_category')['debt'].sum()
purpose_debt_count = case_purpose_data_categorized.groupby('purpose_category')['debt'].count()
purpose_data_debt_percent = purpose_debt_sum / purpose_debt_count * 100


case_purpose = pd.concat([purpose_data_debt_percent, purpose_debt_count, purpose_debt_sum], axis=1)
case_purpose.columns = ['percentage_of_debts','count','sum']
display(case_purpose.sort_values('percentage_of_debts'))


percentage_of_debts  count  sum
purpose_category                                         
операции с автомобилем               7.666903   4239  325
получение образования                7.849571   3962  311
операции с недвижимостью             8.235515  10649  877
проведение свадьбы                   8.904409   2291  204

**Вывод:** 
Наиболее надежная категория - **операции с автомобилем**. Низкий процент должников и большое число заёмщиков говорят о том, что люди стараются вовремя закрывать свои кредитные обязательства в этой области. 

Наибольшее число заемщиков в категории **операции с недвижимостью**. Процент должников средний. Жильё нужно всем, суммы кредитов в этой категории большие, поэтому не все могут справиться. К тому же, если рассматривать ипотеку, сроки кредитования очень долгие и по ним нужна дополнительная информация, чтобы подробнее изучить эту группу.

Наконец, самая ненадежная категория - **проведение свадьбы**. В ней меньше всего заемщиков (но достаточно, чтобы сделать выводы) и самый большой процент должников. Можно сделать вывод, что неадекватная оценка своей платежеспособности происходит под воздействием импульсивных мыслей, что и приводит людей из этой категории к необдуманному решению.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* пропуски обнаружены в столбцах с трудовым стажем (**days_employed**) и доходом (**total_income**).

Возможные причины пропусков в данных: 
 - техническая ошибка: неправильная обработка нулевых данных при заполнении (когда стажа/дохода действительно нет);
 - техническая ошибка при считывании;
 - человоческий фактор (невнимательность при заполнении данных).

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* медиана лучше характеризует выборку из количественных элементов, т.к. она показывает центральное местоположение в последовательности и меньше зависит от резких перепадов значений в выборке. Функия **median( )** упорядочивает выборку и возвращает её центральное значение, если наблюдений нечетное количество и среднее арифметическое двух центральных значений, если количество наблюдений в выборке четное.

### Шаг 4: общий вывод.

Мы провели предварительную обработку данных:
 - **заполнили пропуски** количественных значений в **доходе** и **стаже** медианой по каждой категории типа занятости;
 - обработали **аномальные отрицательные значения** в столбце со **стажем** и **количеством детей**; 
 - в столбце **дохода** заменили **вещественный** тип данных на **целочисленный**;
 - удалили дубликаты;
 - подготовили функции для категоризации данных по **уровню дохода** и **целям** кредита;
 
 После этого приступили к исследованию.

**1) Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Исследование **подтвердило гипотезу** о зависимости между количеством детей и возвратом кредита в срок. Чем больше детей, тем меньше вероятность возврата. К примеру, в группе **без детей** **7.5%** должников, а в группе с **двумя детьми** уже **9.4%**.

**2) Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Эта гипотеза так же **подтвердилась**. Самый большой процент должников **9,7%** - у тех, кто **не был женат/замужем**, что говорит о том, что наличие семьи благоприятно сказывается на вероятности возврата вовремя. У людей **в браке** процент должников заметно ниже -  **7,5%**.

**3) Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

На основе имеющихся данных эта гипотеза **не подтвердилась**. Процент должников по основным группам был примерно одинаковый. Однако, стоит учитывать и простую логику - чем выше доход, тем меньше берут кредиты и тем проще их вернуть в срок.

**4) Как разные цели кредита влияют на его возврат в срок?**

Эта гипотеза **подтвердилась**, самой ненадежной оказалась наиболее импульсивная группа - тех, кто брал кредит на проведение **свадьбы (8.9% должников)**, а самой надежной - **на операции с автомобилем (7.6%)**.

Можно отметить, что во всех проведенных исследованиях, **процент должников по каждой статистически значимой группе не превышал 10 от общего числа заемщиков в своей группе**, резких и аномальных отклонений замечено не было. Это говорит о хорошей статистике возвратов и о надежности данных, на основании которых были проведены исследования.